src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsProjectJsonToPackageReferenceMigrator.cs

```diff
@@ -0,0 +1,89 @@
+{
+    [Export(typeof(IVsProjectJsonToPackageReferenceMigrator))]
+    internal class VsProjectJsonToPackageReferenceMigrator : IVsProjectJsonToPackageReferenceMigrator
+    {
+        private readonly IVsSolutionManager _solutionManager;
+
+        [ImportingConstructor]
+        public VsProjectJsonToPackageReferenceMigrator(IVsSolutionManager solutionManager)
+        {
+            _solutionManager = solutionManager;
+        }
+        public IVsProjectJsonToPackageReferenceMigrateResult MigrateProjectToPackageRef(string projectUniqueName)
+        {
+            ThreadHelper.ThrowIfNotOnUIThread();
+            if(string.IsNullOrEmpty(projectUniqueName))
+            {
+                throw new ArgumentNullException(nameof(projectUniqueName));
+            }
+
+            if(!File.Exists(projectUniqueName))
+            {
+                throw new FileNotFoundException(nameof(projectUniqueName));
+            }
+
+            return  NuGetUIThreadHelper.JoinableTaskFactory.Run(async () => 
+            {
+                await TaskScheduler.Default;
+                return await MigrateProjectToPackageRefAsync(projectUniqueName);
+            });
+            
+        }
+
+        private async Task<IVsProjectJsonToPackageReferenceMigrateResult> MigrateProjectToPackageRefAsync(string projectUniqueName)
+        {            
+            var project = _solutionManager.GetDTEProject(projectUniqueName);
+
+            if (project == null)
+            {
+                throw new InvalidOperationException(string.Format("Project {0} does not exist in the project system cache.", projectUniqueName));
+            }
+
+            var projectSafeName = await EnvDTEProjectUtility.GetCustomUniqueNameAsync(project);
+
+            var nuGetProject = _solutionManager.GetNuGetProject(projectSafeName);
+            try
+            {
+                var legacyPackageRefBasedProject = await
+                    _solutionManager.UpdateNuGetProjectToPackageRef(nuGetProject) as LegacyCSProjPackageReferenceProject;
+
+                var result = new VsProjectJsonToPackageReferenceMigrateResult(await
+                    ProjectJsonToPackageRefMigrator.MigrateAsync(
+                        legacyPackageRefBasedProject,
+                        legacyPackageRefBasedProject.MSBuildProjectPath));
+                
+                if(result.IsSuccess)
+                {
+                    _solutionManager.SaveProject(nuGetProject);
+                }
+
+                return result;
+            }
+            catch(Exception ex)
+            {
+                // reload the project in memory from the file on disk, discarding any changes
+                ReloadProject(project);
+                return new VsProjectJsonToPackageReferenceMigrateResult(ex.Message);
+            }            
+        }
+
+        private void ReloadProject(Project project)
+        {
+            project = _solutionManager.GetDTEProject(project.FullName);
```

Comment:
this will cause any changes in memory to be discarded so it won't prompt the user to save changes or reload project.